### ISSUE TITLE
Audio files not visible in folders on Technisat DigitRadio 580

### ISSUE THREAD

**Issue by cweiske** [gerbera/gerbera#2894]
Since upgrading to Gerbera 1.12.1, my Technisat DigitRadio 580 does not show audio files anymore. When navigating into a folder with audio files only it shows `<Leer>` which means `<Empty>`.

I have transcoding setup for `.ogg` files, but the folders with `.mp3` files only exhibit the same problem.
The radio is showing files when using `minidlna` as UPnP server.

After much testing I managed to get debug logs from the radio with the help of the [Frontier Silicon API](https://github.com/cweiske/fsapi/blob/debug/FSAPI.md#misc):
```
curl "digira/fsapi/SET/netRemote.misc.fsDebug.component?pin=1234&value=14"
curl 'digira/fsapi/SET/netRemote.misc.fsDebug.traceLevel?pin=1234&value=5'
telnet digira 514
```

When browsing a folder with 4 mp3 files on Gerbera it shows:
```
(Thread2): [     87.986738] AVB    (3): Browse for '34203', from 0, max 100 items, server 'f42aa029-7c99-47d2-b6ed-53b9e06cc3a7'
(Thread2): [     87.987678] AVB    (4): Using sort criteria '+upnp:originalTrackNumber'
(Thread2): [     87.988558] AVB    (4): Invoking AVB_REQTYPE_BROWSE request (b0b449e8 b0b449e8)
(Thread2): [     88.087921] AVB    (5): (MS_BrowseResponded) errorCode=0
(Thread2): [     88.099874] AVB    (4): Resource: Not found
(Thread2): [     88.101841] AVB    (4): 2: (0xa200) (68044)  2 Become 1(Single Version)
[...]
(Thread2): [     88.129514] AVB    (3): Browse OK - 4 matches, 4 returned, 4 copied
(Thread2): [     88.129649] AVB    (3): TotalMatches = 4
```

The same on minidlna:
```
(Thread2): [    154.827727] AVB    (3): Browse for '1$7$62', from 0, max 100 items, server '4d696e69-444c-164e-9d41-e2a6051ae345'
(Thread2): [    154.828729] AVB    (4): Using sort criteria '+upnp:originalTrackNumber'
(Thread2): [    154.829665] AVB    (4): Invoking AVB_REQTYPE_BROWSE request (b0c771c0 b0c771c0)
(Thread2): [    154.872768] AVB    (5): (MS_BrowseResponded) errorCode=0
(Thread2): [    154.878909] AVB    (4): Resource: duration=245010, protocolInfo=http-get:*:audio/mpeg:DLNA.ORG_PN=MP3;DLNA.ORG_OP=01;DLNA.
(Thread2): [    154.880154] AVB    (4): Resource: uri=http://192.168.3.5:8200/MediaItems/1601.mp3
(Thread2): [    154.880281] AVB    (4): Resource: protection=
(Thread2): [    154.880428] AVB    (4): 2: (0xa200) (1$7$62$0)  2 Become 1(Single Version)
[...]
(Thread2): [    154.903935] AVB    (3): Browse OK - 4 matches, 4 returned, 4 copied
(Thread2): [    154.904512] AVB    (3): TotalMatches = 4
```

So it detects the single songs in the folder but does not find a resource it could use.

DIDL-Lite from gerbera for a song:
```
<?xml version="1.0" encoding="UTF-8"?>
<DIDL-Lite xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns:sec="http://www.sec.co.kr/dlna">
  <item id="68044" parentID="34203" restricted="1">
    <dc:title>2 Become 1(Single Version)</dc:title>
    <upnp:class>object.item.audioItem.musicTrack</upnp:class>
    <dc:date>1996-01-01</dc:date>
    <dc:description>Spice Girls, 2 Become 1, 2 Become 1(Single Version), 1996, Pop</dc:description>
    <upnp:album>2 Become 1</upnp:album>
    <upnp:artist>Spice Girls</upnp:artist>
    <upnp:date>1996</upnp:date>
    <upnp:genre>Pop</upnp:genre>
    <upnp:originalTrackNumber>1</upnp:originalTrackNumber>
    <upnp:albumArtURI xmlns:dlna="urn:schemas-dlna-org:metadata-1-0" dlna:profileID="JPEG_TN">http://mediatomb.home.cweiske.de:49152/content/media/object_id/68044/res_id/1/rct/aa/ext/file.jpg</upnp:albumArtURI>
    <res size="4229120" duration="0:04:05.002" bitrate="16384" sampleFrequency="44100" nrAudioChannels="2" protocolInfo="http-get:*:audio/mpeg:DLNA.ORG_PN=MP3;DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=01700000000000000000000000000000">http://mediatomb.home.cweiske.de:49152/content/media/object_id/68044/res_id/0/group/default/ext/file.mp3</res>
    <res protocolInfo="http-get:*:image/jpeg:DLNA.ORG_PN=JPEG_TN;DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=00800000000000000000000000000000">http://mediatomb.home.cweiske.de:49152/content/media/object_id/68044/res_id/1/rct/aa/ext/file.jpg</res>
  </item>
  [...]
```

The same from minidlna:
```
<?xml version="1.0" encoding="UTF-8"?>
<DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/">
  <item id="1$7$62$0" parentID="1$7$62" restricted="1">
    <dc:title>2 Become 1(Single Version)</dc:title>
    <upnp:class>object.item.audioItem.musicTrack</upnp:class>
    <res duration="0:04:05.010" protocolInfo="http-get:*:audio/mpeg:DLNA.ORG_PN=MP3;DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=01700000000000000000000000000000">http://192.168.3.5:8200/MediaItems/1601.mp3</res>
  </item>
  [...]
```

I'm currently stuck with that. I also do not remember which gerbera version I had before.
Setting client flags `IRADIO|STRICTXML` did not help.

What could I try?

**Comment by cweiske** [comment]
Luckily I had tagged the old version I was using, and it was commit 77b2d3bb74b0797fb5bc88c5cc902230cce6f5c1.

Comparing the DIDL-Lite XML from the old version to the new one showed - apart from the different order or XML tag attributes - that Gerbera now uses the virtual URI in UPnP responses. The old version only used the IP address:

Old:
```
http://192.168.3.3:49152/content/media/object_id/736/res_id/0/ext/file.ogg
```

New:
```
http://mediatomb.home.cweiske.de:49152/content/media/object_id/51123/res_id/0/group/default/ext/file.ogg
```

After changing `<virtualURL>` to contain an IP address instead of a hostname, the radio lists and plays the songs again.

I would really like to keep the domain-based `virtualURL` for the web interface, though.

**Comment by cweiske** [comment]
The question is if this behavior is mandated by the UPnP/DLNA spec, or if it's just a bug/missing feature in the frontier smart firmware.

The radio itself has a DNS server configured via DHCP, and that one can resolve the hostname.

**Comment by KarlStraussberger** [comment]
In fact we changed the behaviour in 1.12.1 to allow setting browse contents in a container. It depends on the device to resolve dns names in content. As a consequence of your and others issue, there is a new setting in head (https://docs.gerbera.io/en/latest/config-server.html#virtualurl)

**Comment by cweiske** [comment]
Thank you for the information!
Someone else had a similar issue with `virtualURL` at #2835, and your fix was #2836.
I'll use `externalURI` now in my config as well.
